message_ix/macro.py

```diff
@@ -1,61 +1,494 @@
+        gdp = self.data['gdp_calibrate']
+        gdp0 = gdp.iloc[gdp.index.isin([self.init_year], level='year')]
+        self.data['gdp0'] = gdp0
+        return self.data['gdp0']
+
+    @lru_cache()
+    def _k0(self):
+        kgdp = self.data['kgdp']
+        gdp0 = self._gdp0()
+        self.data['k0'] = kgdp * gdp0
+        return self.data['k0']
+
+    @lru_cache()
+    def _total_cost(self):
+        # read from scenario
+        idx = ['node', 'year']
+        model_cost = self._clean_model_data(self.s.var('COST_NODAL_NET'))
+        model_cost.rename(columns={'lvl': 'value'}, inplace=True)
+        model_cost = model_cost[idx + ['value']]
+        # TODO: in the R code, this value is divided by 1000
+        # do we need to do that here?!!?
+        model_cost['value'] /= 1e3
+        # get data provided in init year from data
+        cost_ref = self.data['cost_ref'].reset_index()
+        cost_ref['year'] = self.init_year
+        # combine into one value
+        total_cost = pd.concat([cost_ref, model_cost]).set_index(idx)['value']
+        if total_cost.isnull().any():
+            raise RuntimeError('NaN values found in total_cost calculation')
+        self.data['total_cost'] = total_cost
+        return total_cost
+
+    @lru_cache()
+    def _price(self):
+        # read from scenario
+        idx = ['node', 'sector', 'year']
+        model_price = self._clean_model_data(
+            self.s.var('PRICE_COMMODITY', filters={'level': 'useful'})
+        )
+        if np.isclose(model_price['lvl'], 0).any():
+            # TODO: this needs a test..
+            msg = '0-price found in MESSAGE variable PRICE_COMMODITY'
+            raise RuntimeError(msg)
+        model_price.rename(columns={'lvl': 'value', 'commodity': 'sector'},
+                           inplace=True)
+        model_price = model_price[idx + ['value']]
+
+        # get data provided in init year from data
+        price_ref = self.data['price_ref'].reset_index()
+        price_ref['year'] = self.init_year
+        # combine into one value
+        price = pd.concat([price_ref, model_price]).set_index(idx)['value']
+        if price.isnull().any():
+            raise RuntimeError('NaN values found in price calculation')
+        self.data['price'] = price
+        return price
+
+    @lru_cache()
+    def _demand(self):
+        # read from scenario
+        idx = ['node', 'sector', 'year']
+        model_demand = self._clean_model_data(
+            self.s.var('DEMAND', filters={'level': 'useful'})
+        )
+        model_demand.rename(columns={'lvl': 'value', 'commodity': 'sector'},
+                            inplace=True)
+        model_demand = model_demand[idx + ['value']]
+        # get data provided in init year from data
+        demand_ref = self.data['demand_ref'].reset_index()
+        demand_ref['year'] = self.init_year
+        # combine into one value
+        demand = pd.concat([demand_ref, model_demand]).set_index(idx)['value']
```

Comment:
A minor suggestion to turn off the deprecation warnings:
```
 demand = pd.concat([demand_ref, model_demand], sort=True).set_index(idx)['value']

Reply:
See https://github.com/iiasa/message_ix/pull/223/files#r392681157